### ISSUE TITLE
Add accessible labels

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/friction#49]
After 11/28/17 a11y meeting, we would like to add accessible labels to  navigable components. We need labels for the book and the magnified view of the book since both of those are focusable. Other buttons have common code labels. This sim was next on the docket for publication so it would be nice to have this done before next week.

Maybe "Chemistry Book" and "Chemistry Book, atomic view"?

@terracoda @zepumph @emily-phet thoughts?

**Comment by terracoda** [comment]
@jessegreenberg, I have a design document for this one, too.
https://docs.google.com/document/d/1XXI7SD-ASJwivsBaHgeOZmMcJFFN_s3FU48-syO-G0o/edit#

I think "Atomic view" is likely better than "Magnified view" which is what I had originally. I'll update the document.

I'm still not sure about the HTML draft structure, but we can look at that later.

**Comment by terracoda** [comment]
@jessegreenberg, I also like the labels without the verb, "Move". The verb would ideally go in the on-demand help text when the HTML is sorted out.

**Comment by zepumph** [comment]
@jessegreenberg please review this one too.

**Comment by terracoda** [comment]
@jessegreenberg, and @zepumph re my comment
> I'm still not sure about the HTML draft structure, but we can look at that later.

I need to retract that statement. You can't name an object without the HTML. I apologize for my temporary lapse in reasoning.

Because the Chemistry book is a custom element, giving it a label that can be heard by AT is not as simple as adding a `label` element to a typical form control.

The HTML and labeling technique will be closer to the structure of a balloon in BASE, but without the button to launch the custom web application.

**Comment by jessegreenberg** [comment]
@zepumph changes look great, thanks. I am surprised, when I tried to use the sim with NVDA in FFQ I don't hear the labels when I tab to the books. When I change the HTML to
```html
<div role="application">
  <div>‪Chemistry‬ book</div>
</div>
```
NVDA does read the label.

**Comment by jessegreenberg** [comment]
I just committed that change. Otherwise I think this can be closed, but back to @zepumph in case there is anything else.

**Comment by terracoda** [comment]
I was going to ask for the HTML. You'll need a labeling technique:

```html
<div role="application" aria-label="id-chem-book-label">
  <div id="chem-book-label">‪Chemistry‬ book</div>
</div>```

Or something like that.

**Comment by terracoda** [comment]
Oops, in my haste that is incorrect code.

**Comment by terracoda** [comment]
Adding `aria-labelledby` should make the name available as an _accessible name_ for the Web Application in the Accessibility Tree, but I haven't tested it.

**Comment by terracoda** [comment]
@jessegreenberg or @zepumph could you try the `aria-labelledby` attribute? and see if that makes the name audible when using a screen reader?

**Comment by jessegreenberg** [comment]
Here is the full HTML
```html
<div tabindex="-1" id="container-264-325-341" role="application">
    <div id="264-325-341" tabindex="0">‪Chemistry‬ book</div>
</div>

Since the inner div is focusable, NVDA is making "Chemistry Book" is the accessible name for the focused element. What does aria-labelledby do if the div with role=application is not focusable?

**Comment by terracoda** [comment]
I'll test again in VO. I didn't hear the name, only "Web Application" when I last listened.

**Comment by jessegreenberg** [comment]
Did you try with https://github.com/phetsims/friction/commit/cd9d735412ebd599b3e36faad05c8b25c2d2dfd2 (change made 11 min ago)?

**Comment by terracoda** [comment]
No, and I read your comment incorrectly :-) Just testing again now.

**Comment by terracoda** [comment]
@jessegreenberg, I  hear "Chemistry book, group, web application" for the book, but only hear "web application" for the Atomic view of the book.

Do you have the same code on the atomic view?
```html
<div tabindex="-1" id="container-id-number" role="application">
    <div id="id-number" tabindex="0">‪Chemistry‬ book, atomic view</div>
</div>
```

**Comment by terracoda** [comment]
@jessegreenberg, you are correct the name of the focusable `div` is being made available as the accessible name. Nice work.

**Comment by jessegreenberg** [comment]
Ah, good catch it doesn't look like it! For the atomic view, role=application is on the inner div. But for my version of VO, I am hearing something different :(

For this html:
```html
<div tabindex="-1" id="container-264-325-341" role="application">
    <div id="264-325-341" tabindex="0">‪Chemistry‬ book</div>
</div>
```
I hear "Chemisty Book,group".

For this html:
```html
<div tabindex="-1" id="container-264-325-358-359-402-404">
    <div id="264-325-358-359-402-404" tabindex="0" role="application">‪Chemistry‬ book, Atomic view</div>
</div>
```
I hear "Chemistry Book,  Atomic View, Web Application".

**Comment by jessegreenberg** [comment]
Now NVDA is failing to read the accessible name for either. So I am not sure how to proceed. @terracoda what would you recommend?

**Comment by terracoda** [comment]
@jessegreenberg, shouldn't the role application be implemented the same for both objects?, i.e., on the outer div?

In my VO I only hear the name on the book, not on the atomic view.

**Comment by jessegreenberg** [comment]
Yes it should, but the difference shows how AT treat them differently. And my version of VO (10.0.1, admittedly an older version now!) prefers the HTML for the atomic view.  But since NVDA doesn't read either in FF Quantum I was wondering if something totally different was required.

**Comment by terracoda** [comment]
Strange my version of VO just updated, prefers the Book.
I'm not sure if we should rely on FF Quantum. Is it possible to try with an older version of FF? Do you have a machine available to do that?

I checked the HTML in BASE and it seems very similar, if not the same as the Book in Friction.

**Comment by jessegreenberg** [comment]
VO in Safari 10 and NVDA in FF Quantum both read labels correctly with this HTML, @terracoda thoughts?
```html

**Comment by jessegreenberg** [comment]
I will look into getting an earlier  version of FF.

**Comment by terracoda** [comment]
@jessegreenberg, I was going to suggest an explicit labeling technique, and was just re-reading specification suggestions for role application.

https://www.w3.org/TR/wai-aria-1.1/#application

**Comment by jessegreenberg** [comment]
Sounds good. Since https://github.com/phetsims/friction/issues/49#issuecomment-347996129 seems to work OK locally I am inclined to go with it. But @terracoda let us know if you would recommend something else after reading the specs. Changes committed and can be reviewed on master.

**Comment by terracoda** [comment]
I was intrigued by the suggestion in the spec that one element with role application can have two focusable items.

I'm a little uncomfortable with an empty div with aria-label. I have a feeling that aria-labelledby might be better, but we can stick with this if it is valid code.

**Comment by jessegreenberg** [comment]
Thanks for clarifying, back to aria-labelledby as you originally suggested @terracoda, it sounds great in Safari 10.0.1 and FF Quantum. I think that is all then, @zepumph could you please re-review https://github.com/phetsims/friction/commit/edefa5b9d3aa8ae20e3b4648caf2299d08935be8 and test with your NVDA?

**Comment by zepumph** [comment]
On firefox NVDA sounds nice, on Chrome, for the macro book I hear "Chemistry Book, Chemistry Book Section." For the magnified view I hear "Chemistry Book, Atomic View, Chemistry Book, Atomic View Section." I'm not sure what "section" is, but since this isn't a supported pairing on not worried about it. I think we are good to go. Back to you @jessegreenberg to review and close.

**Comment by terracoda** [comment]
@zepumph, sounds like Chrome is rendering the role application as a "section" rather than a "web application". I have a feeling the "section" utterance could change over time.

**Comment by jessegreenberg** [comment]
Thanks @zepumph, great. I think @terracoda is correct. Closing this issue.
